### DISCUSSION TITLE
USD to Maya attribute connections

### DISCUSSION THREAD

**Discussion by nomad3d** [Autodesk/maya-usd#1881] [Q&A]
Hi, 
I'm trying to get an attribute connection out of the USD proxy shape to a Maya native object. 
To be more specific:
I try to create a Maya native camera, which gets its input attributes from the USD proxy shape. 
The proxyAccessor has a couple of functions for getting and creating attributes, but no matter what I do, the plug doesn't  seem to work.
I only get the "AP__" attribute on the proxyShape  with a default 0.0 value, but when I "getOrCreate" the transform matrix, the new matrix attribute gets instantly updated correctly. I couldn't find any documentation regarding attribute connections and most of the examples deal with with the xform linking between USD and Maya objects. To me it looks like that the UFE doesn't even recognise the plug.
Is this even possible with the current version?

**Comment by kxl-adsk** [comment]
Hi @nomad3d. Can you create a simple scene/script to reproduce the issue? It could be that specific attribute type is not handled by the converter...but I would prefer to see the setup than speculate. 

Cheers,
KxL

**Comment by nomad3d** [comment]
Hi @kxl-adsk , In the attached sample you can see a successful connection of the xforms and and the "broken" focalLength attribute on the proxyShape.
Thanks for looking into it!

[usdTest.zip](https://github.com/Autodesk/maya-usd/files/7667899/usdTest.zip)

**Comment by kxl-adsk** [comment]
You can turn on extra debugging traces for proxy accessor by enabling `USDMAYA_PROXYACCESSOR` traces. Here is how you can do it from python:
```
from pxr import Tf

Tf.Debug.SetDebugSymbolsByName("USDMAYA_PROXYACCESSOR", 1)
```

Then you will notice a message in the console:
```
...
Skipped attribute, no valid converter found for '/camMatchmove_MP1.focalLength'
...
```

And this is now embarrassing, we haven't implemented converted for `float` yet - https://github.com/Autodesk/maya-usd/blob/dev/lib/mayaUsd/utils/converter.cpp#L38. There is a bit of explanation for it, we wanted to show patterns for different possible type patterns and `float` would simply follow `int` pattern.

**Comment by kxl-adsk** [comment]
Here is a code change that adds support for floats
```
diff --git a/lib/mayaUsd/utils/converter.cpp b/lib/mayaUsd/utils/converter.cpp
index 6afee9166..ccd9fe09c 100644
--- a/lib/mayaUsd/utils/converter.cpp
+++ b/lib/mayaUsd/utils/converter.cpp
@@ -190,6 +190,26 @@ template <> struct MakeMayaSimpleData<int> : public std::true_type
     }
 };
 
+//! \brief  Type trait for Maya's float type providing get and set methods for data handle and
+//! plugs.
+template <> struct MakeMayaSimpleData<float> : public std::true_type
+{
+    using Type = float;
+    enum
+    {
+        kNumericType = MFnNumericData::kFloat
+    };
+
+    static void get(const MDataHandle& handle, Type& value) { value = handle.asFloat(); }
+
+    static void set(MDataHandle& handle, const Type& value) { handle.setFloat(value); }
+
+    static void set(const MPlug& plug, MDGModifier& dst, const Type& value)
+    {
+        dst.newPlugValueFloat(plug, value);
+    }
+};
+
 //---------------------------------------------------------------------------------
 //! \brief  Type trait declaration for complex data types. Each specialized type will inherit
 //! from std::true_type
@@ -1008,6 +1028,7 @@ struct Converter::GenerateConverters
 
         createConverter<bool, bool>(converters, SdfValueTypeNames->Bool);
         createConverter<int, int32_t>(converters, SdfValueTypeNames->Int);
+        createConverter<float, float>(converters, SdfValueTypeNames->Float);
 
         createConverter<MString, std::string>(converters, SdfValueTypeNames->String);
         createConverter<float3, GfVec3f>(converters, SdfValueTypeNames->Float3);
```

I hope this unblocks you. For the final change, it will require a unit test...and we should right away add double support as well before we can open a PR.

**Reply by kxl-adsk** [reply]
PR has been opened to add support for float and double - https://github.com/Autodesk/maya-usd/pull/1892

**Reply by nomad3d** [reply]
Great, Thank you!

**Comment by nomad3d** [comment]
Hi, @kxl-adsk,

I've just noticed that float2 attributes still don't work:

`Skipped attribute, no valid converter found for '/cameras/camera1.clippingRange'`

We are on plug in version 0.17.0

Thanks for looking into it.

**Reply by kxl-adsk** [reply]
Hi @nomad3d, I no longer work at Autodesk, so can’t help you here. 

If it’s blocking you, adding float2 should be pretty straightforward if you follow float3 example. 

Cheers,
KxL
